fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/FDBIndexOutputTest.java

```diff
@@ -53,4 +65,50 @@ public void testWriteByte() throws Exception {
         assertEquals(1, directory.getFDBLuceneFileReference(FILE_NAME).getSize());
     }
 
+    @Test
+    public void testCopyBytesPipeline() throws Exception {
+        FDBIndexOutput output = new FDBIndexOutput(FILE_NAME, directory);
+        output.writeBytes(BLOCK_ARRAY_100, BLOCK_ARRAY_100.length);
+        output.close();
+        assertEquals(BLOCK_ARRAY_100.length, directory.getFDBLuceneFileReference(FILE_NAME).getSize());
+        IndexInput blocks = directory.openInput(FILE_NAME, IOContext.READONCE);
+        output = new FDBIndexOutput(FILE_NAME_TWO, directory);
+        output.copyBytes(blocks, blocks.length());
+        output.close();
+        assertEquals(BLOCK_ARRAY_100.length, directory.getFDBLuceneFileReference(FILE_NAME_TWO).getSize());
+    }
+
+    @Test
+    public void testCopyBytesReadAheadPipelineShortBuffer() throws Exception {
+        FDBIndexOutput output = new FDBIndexOutput(FILE_NAME, directory);
+        output.writeBytes(BLOCK_ARRAY_100, BLOCK_ARRAY_100.length);
+        output.close();
+        assertEquals(BLOCK_ARRAY_100.length, directory.getFDBLuceneFileReference(FILE_NAME).getSize());
+        IndexInput blocks = directory.openChecksumInput(FILE_NAME, IOContext.READONCE);
+        output = new FDBIndexOutput(FILE_NAME_TWO, directory);
+        output.setNumberOfBytesToPrefetch((PrefetchableBufferedChecksumIndexInput) blocks, 2 * 16 * 1024);
+        assertEquals(2, directory.getBlockCache().asMap().size());
+    }
+
+    @Test
+    public void testCopyBytesReadAheadPipelineGreaterThanBuffer() throws Exception {
+        FDBIndexOutput output = new FDBIndexOutput(FILE_NAME, directory);
+        output.writeBytes(BLOCK_ARRAY_100, BLOCK_ARRAY_100.length);
+        output.close();
+        assertEquals(BLOCK_ARRAY_100.length, directory.getFDBLuceneFileReference(FILE_NAME).getSize());
+        IndexInput blocks = directory.openChecksumInput(FILE_NAME, IOContext.READONCE);
+        output = new FDBIndexOutput(FILE_NAME_TWO, directory);
+        output.setNumberOfBytesToPrefetch((PrefetchableBufferedChecksumIndexInput) blocks, 50 * 16 * 1024);
+        assertEquals(10, directory.getBlockCache().asMap().size());
+        byte[] fooey = new byte[16 * 1024];
+        blocks.readBytes(fooey, 0, fooey.length); // 1 block (refresh)
+        assertEquals(11, directory.getBlockCache().asMap().size()); // Maintains Minimum Buffer
+        blocks.readBytes(fooey, 0, fooey.length); // next block, no fetch
+        blocks.readBytes(fooey, 0, fooey.length); // another block, no fetch
+        assertEquals(13, directory.getBlockCache().asMap().size()); // Waits until buffer shrinks
+        fooey = new byte[10 * 16 * 1024];
+        blocks.readBytes(fooey, 0, fooey.length); // another block, no fetch
+        assertEquals(25, directory.getBlockCache().asMap().size()); // Waits until buffer shrinks
```

Comment:
I am not sure I understand, why this is 25 instead of 23? it just reads another 10 blocks after the 13 asserted at line 108 right?

Reply:
Good catch, fixing.